aapcs64-morello/aapcs64-morello.rst

```diff
@@ -521,22 +526,22 @@ The differences in language bindings used for AAPCS64 and AAPCS64-cap are descri
-    +------------------------------+------------------------------------------------------------------------------------------+
-    |                              | If the callee is variadic in AAPCS64-cap, the Anonymous Arguments Memory area is         |
-    |                              | allocated in memory with a size of at least 16 * (number of Anonymous arguments) bytes.  |
-    |                              | The Anonymous Arguments Memory area is 16-byte aligned. The Anonymous Arguments          |
-    |                              | Capability points to the Anonymous Arguments Memory area and is copied to C9. If there   |
-    | A.4                          | are no Anonymous arguments passed, C9 is set to NULL.                                    |
+    |                              | If any parameters are memory passed in AAPCS64-cap, the Arguments Memory Area is         |
+    |                              | allocated.                                                                               |
+    |                              | The Arguments Memory area is 16-byte aligned. The Arguments Capability has the           |
+    |                              | bounds of the Arguments Memory area and Arguments Memory area and the address of the     |
+    |                              | first memory passed argument. The Arguments capability is copied to C9. If there are no  |
+    | A.3                          | Anonymous arguments and the callee is variadic, C9 is set to NULL. See notes.            |
     +------------------------------+------------------------------------------------------------------------------------------+
-    |                              | The Anonymous Arguments Index (AArgsIdx) is set to zero.                                 |
-    |                              |                                                                                          |
-    | A.5                          |                                                                                          |
+    |                              | The next memory argument address (NMAA) is set to the value of C9 in AAPCS64-cap or the  |
+    | A.4                          | current stack-pointer value (SP) otherwise.                                              |
     +------------------------------+------------------------------------------------------------------------------------------+
 
 
+.. note::
+   The Arguments Capability is a capability that has the bounds of the arguments memory area and its address will be that of the first memory passed argument. The address of the Arguments Capability is 16-byte aligned. The capability offset of the Arguments Capability may not be zero. The capability offset is such that the first 16-byte aligned location after the last memory passed argument will be larger or equal than capability limit of the Arguments Capability.
```

Comment:
"larger than or equal to the capability limit..."

I wonder if this paragraph shouldn't start by explaining how the Arguments Memory Area is defined. I don't really understand how it all fits in terms of bounds representability right now.

Reply:
I agree about the precision issues, but also it'd be very useful to include a diagram here. I don't know if we have the infrastructure for that in this document, but it would really help.

I'm a bit confused about the relationship between the offset and bounds at various points. For example, here the "address will be that of the first memory passed argument", but also its offset may not be zero. However, won't the first memory-passed argument be allocated to the first slot in the arguments memory area, and therefore require an offset of zero?

It's likely that I've just misunderstood something obvious!

Reply:
Thanks! I'll try to clarify this in some way and look at adding a diagram. We need to add some padding at the start of the arguments area to make the length representable. So the 'offset' (maybe this should be 'capability offset'?) is just the size of the padding, and the first argument (and first slot) is located at the address (capability value in Morello terms) of C9.

We don't say how the (capability) offset is computed and leave that up to the implementation. In practice the way to do this would be to run the algorithm once to get the number of bytes that we need for argument passing and compute from that the minimal amount of padding required to make everything representable. With the current wording we can have multiple legal values of offset, but just one being minimal/optimal.

Hopefully this clarifies things?

Reply:
Ok, that does clarify things, but couldn't the offset still be zero in the (common) case where the area is precisely representable without padding?

Reply:
At the moment any sane implementation should give a zero offset in this case. We could add the additional constraint that the offset needs to be minimal, which would then require a zero offset. Any thoughts?